### ISSUE TITLE
Stack overflow in SCEV when constructing SCEV expression recursively.

### ISSUE THREAD

**Issue by dbakunevich** [llvm/llvm-project#55333]
This bug was found by Azul FuzzGen fuzzer for CPP.

Godbolt link: https://godbolt.org/z/KeEWv8WxK

Run clang -O3 (x86_64) on the test
[crash_3.cpp.txt](https://github.com/llvm/llvm-project/files/8646897/crash_3.cpp.txt)

The bug was found on all versions from 6 to 15.

Failure looks like:

```
./gen_04441.cpp:576:29: warning: unsequenced modification and access to 'int8_tVar_59' [-Wunsequenced]
if ( (int8_t)( min((int8_t)(++int8_tVar_59), (int8_t)((int8_t)int8_tVar_59)) ) ) {                                                                                                                                                                                       ^                                 ~~~~~~~~~~~~
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:                                                                                                                                                                                                                                  0.      Program arguments: /var/tmp/clang_15.0.0/RA_3779/bin/clang-15 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj --mrelax-relocations -disable-free -clear-ast-before-backend -main-file-name gen_04441.cpp -mrelocation-model static -mframe-pointer=none -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debugger-tuning=gdb -fcoverage-compilation-dir=/home/tatyana/work/cpp/run_long_clang_15.0.0_fd_3779_9_1/tests -resource-dir /var/tmp/clang_15.0.0/RA_3779/lib/clang/15.0.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/x86_64-linux-gnu/c++/7.5.0 -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../include/c++/7.5.0/backward -internal-isystem /var/tmp/clang_15.0.0/RA_3779/lib/clang/15.0.0/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -O3 -fdeprecated-macro -fdebug-compilation-dir=/home/tatyana/work/cpp/run_long_clang_15.0.0_fd_3779_9_1/tests -ferror-limit 19 -fgnuc-version=4.2.1 -fcxx-exceptions -fexceptions -vectorize-loops -vectorize-slp -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/gen_04441-07f4ca.o -x c++ ./gen_04441.cpp                                                                                                                                                       1.      <eof> parser at end of file
2.      Optimizer
  #0 0x00007f4a854fbc2c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Support/Unix/Signals.inc:569:0
  #1 0x00007f4a854f9c24 llvm::sys::RunSignalHandlers() /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Support/Signals.cpp:103:0
  #2 0x00007f4a854fa393 SignalHandler(int) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Support/Unix/Signals.inc:397:0
  #3 0x00007f4a91642980 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12980)
  #4 0x00007f4a853e4817 llvm::hashing::detail::hash_combine_recursive_helper::hash_combine_recursive_helper() /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/ADT/Hashing.h:510:0                                           #5 0x00007f4a853e4817 llvm::hash_code llvm::hash_combine<unsigned int, unsigned long>(unsigned int const&, unsigned long const&) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/ADT/Hashing.h:607:0
  #6 0x00007f4a853e4817 llvm::hash_value(llvm::APInt const&) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Support/APInt.cpp:580:0                                                                                                 #7 0x00007f4a853e4939 llvm::DenseMapInfo<llvm::APInt, void>::getHashValue(llvm::APInt const&) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Support/APInt.cpp:589:0
  #8 0x00007f4a85567777 llvm::DenseMapInfo<llvm::APInt, void>::isEqual(llvm::APInt const&, llvm::APInt const&) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/ADT/APInt.h:2297:0                                           #9 0x00007f4a85567777 bool llvm::DenseMapBase<llvm::DenseMap<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> >, llvm::DenseMapInfo<llvm::APInt, void>, llvm::detail::DenseMapPair<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> > > >, llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> >, llvm::DenseMapInfo<llvm::APInt, void>, llvm::detail::DenseMapPair<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> > > >::LookupBucketFor<llvm::APInt>(llvm::APInt const&, llvm::detail::DenseMapPair<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> > > const*&) const /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/ADT/DenseMap.h:633:0                                                                                                                     #10 0x00007f4a8556e6ec llvm::DenseMapBase<llvm::DenseMap<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> >, llvm::DenseMapInfo<llvm::APInt, void>, llvm::detail::DenseMapPair<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> > > >, llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> >, llvm::DenseMapInfo<llvm::APInt, void>, llvm::detail::DenseMapPair<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> > > >::FindAndConstruct(llvm::APInt const&) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/ADT/DenseMap.h:325:0
 #11 0x00007f4a8556e6ec llvm::DenseMapBase<llvm::DenseMap<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> >, llvm::DenseMapInfo<llvm::APInt, void>, llvm::detail::DenseMapPair<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> > > >, llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> >, llvm::DenseMapInfo<llvm::APInt, void>, llvm::detail::DenseMapPair<llvm::APInt, std::unique_ptr<llvm::ConstantInt, std::default_delete<llvm::ConstantInt> > > >::operator[](llvm::APInt const&) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/ADT/DenseMap.h:332:0
 #12 0x00007f4a8556e6ec llvm::ConstantInt::get(llvm::LLVMContext&, llvm::APInt const&) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/IR/Constants.cpp:917:0
 #13 0x00007f4a86d44a00 llvm::APInt::~APInt() /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/ADT/APInt.h:168:0
 #14 0x00007f4a86d44a00 llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7335:0
 #15 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0                                                                                  #16 0x00007f4a86d460ef llvm::ScalarEvolution::SCEVCallbackVH::SCEVCallbackVH(llvm::Value*, llvm::ScalarEvolution*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:12747:0
 #17 0x00007f4a86d460ef llvm::ScalarEvolution::getSCEV(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:4334:0
 #18 0x00007f4a86d44a3d llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7336:0
 #19 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0
 #20 0x00007f4a86d460ef llvm::ScalarEvolution::SCEVCallbackVH::SCEVCallbackVH(llvm::Value*, llvm::ScalarEvolution*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:12747:0
 #21 0x00007f4a86d460ef llvm::ScalarEvolution::getSCEV(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:4334:0
 #22 0x00007f4a86d44a3d llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7336:0
 #23 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0
 #24 0x00007f4a86d460ef llvm::ScalarEvolution::SCEVCallbackVH::SCEVCallbackVH(llvm::Value*, llvm::ScalarEvolution*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:12747:0
 #25 0x00007f4a86d460ef llvm::ScalarEvolution::getSCEV(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:4334:0
 #26 0x00007f4a86d44a3d llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7336:0
 #27 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0
 #28 0x00007f4a86d460ef llvm::ScalarEvolution::SCEVCallbackVH::SCEVCallbackVH(llvm::Value*, llvm::ScalarEvolution*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:12747:0
 #29 0x00007f4a86d460ef llvm::ScalarEvolution::getSCEV(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:4334:0
 #30 0x00007f4a86d44a3d llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7336:0
 #31 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0
 #32 0x00007f4a86d460ef llvm::ScalarEvolution::SCEVCallbackVH::SCEVCallbackVH(llvm::Value*, llvm::ScalarEvolution*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:12747:0
 #33 0x00007f4a86d460ef llvm::ScalarEvolution::getSCEV(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:4334:0
 #34 0x00007f4a86d44a3d llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7336:0
 #35 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0
 #36 0x00007f4a86d460ef llvm::ScalarEvolution::SCEVCallbackVH::SCEVCallbackVH(llvm::Value*, llvm::ScalarEvolution*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:12747:0
 #37 0x00007f4a86d460ef llvm::ScalarEvolution::getSCEV(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:4334:0
 #38 0x00007f4a86d44a3d llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7336:0
 #39 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0
 #40 0x00007f4a86d460ef llvm::ScalarEvolution::SCEVCallbackVH::SCEVCallbackVH(llvm::Value*, llvm::ScalarEvolution*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:12747:0
 #41 0x00007f4a86d460ef llvm::ScalarEvolution::getSCEV(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:4334:0
 #42 0x00007f4a86d44a3d llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7336:0
 #43 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0
 #44 0x00007f4a86d460ef llvm::ScalarEvolution::SCEVCallbackVH::SCEVCallbackVH(llvm::Value*, llvm::ScalarEvolution*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:12747:0
 #45 0x00007f4a86d460ef llvm::ScalarEvolution::getSCEV(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:4334:0
 #46 0x00007f4a86d44a3d llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7336:0
 #47 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0
....................
#506 0x00007f4a86d44a3d llvm::ScalarEvolution::createSCEV(llvm::Value*) (.part.1961) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:7336:0
#507 0x00007f4a86d460ef llvm::CallbackVH::CallbackVH(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/include/llvm/IR/ValueHandle.h:396:0
#508 0x00007f4a86d460ef llvm::ScalarEvolution::SCEVCallbackVH::SCEVCallbackVH(llvm::Value*, llvm::ScalarEvolution*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:12747:0
#509 0x00007f4a86d460ef llvm::ScalarEvolution::getSCEV(llvm::Value*) /home/dbakunevich/work/llvm-project/llvm-project/llvm/lib/Analysis/ScalarEvolution.cpp:4334:0
clang-15: error: unable to execute command: Segmentation fault (core dumped)
clang-15: error: clang frontend command failed due to signal (use -v to see invocation)
```

**Comment by fhahn** [comment]
This looks like a duplicate of https://github.com/llvm/llvm-project/issues/44546. SCEV overflows the stack due to recursively constructing expressions.

Fixed by the candidate patch for #44546  https://reviews.llvm.org/D114650

**Comment by fhahn** [comment]
This doesn't crash on Godbolt any longer: https://godbolt.org/z/77T9rcKEE (It requires adding a newline to re-trigger compiling with the new compiler build I think)
